src/hathor.c

```diff
@@ -92,64 +68,111 @@ void compress_public_key(unsigned char *value) {
     value[0] = ((value[64] & 1) ? 0x03 : 0x02);
 }
 
-int encode_base58(const unsigned char *in, size_t inlen, unsigned char *out, size_t outlen) {
-    unsigned char buffer[B58_MAX_INPUT_SIZE * 138 / 100 + 1] = {0};
-    size_t i = 0, j;
-    size_t startAt, stopAt;
-    size_t zeroCount = 0;
-    size_t outputSize;
+void pubkey_hash_to_address(uint8_t *hash, uint8_t *out) {
+    unsigned char checksum_buffer[32];
+    // prepend version
+    out[0] = P2PKH_VERSION_BYTE;
+    os_memmove(out+1, hash, 20);
+    // sha256d of above and get first 4 bytes (checksum)
+    sha256d(out, 21, checksum_buffer);
+    os_memmove(out+21, checksum_buffer, 4);
+}
 
-    if (inlen > B58_MAX_INPUT_SIZE) {
-        return -1;
-    }
+void pubkey_to_address(cx_ecfp_public_key_t *public_key, uint8_t *out) {
+    unsigned char hash_buffer[20];
+    // get compressed pubkey
+    compress_public_key(public_key->W);
+    // get hash160
+    hash160(public_key->W, 33, hash_buffer);
+    pubkey_hash_to_address(hash_buffer, out);
+}
 
-    while ((zeroCount < inlen) && (in[zeroCount] == 0)) {
-        ++zeroCount;
+/**
+ * Validates that a script has the format of P2PKH. Throws an exception if doesn't.
+ * P2PKH scripts have the format:
+ *   [OP_DUP, OP_HASH160, pubkey_hash_len, pubkey_hash, OP_EQUALVERIFY, OP_CHECKSIG]
+ * Considering that pubkey hashes have 20 bytes and the values of the opcodes:
+ *   [0x76, 0xA9, 20, pubkey_hash, 0x88, 0xAC]
+ */
+void validate_p2pkh_script(uint8_t *in) {
+    uint8_t p2pkh[] = {0x76, 0xA9, 20, 0x88, 0xAC};
+    if (os_memcmp(p2pkh, in, 3) != 0 || os_memcmp(p2pkh + 3, in + 23, 2) !=0) {
+        THROW(SW_INVALID_PARAM);
     }
+}
 
-    outputSize = (inlen - zeroCount) * 138 / 100 + 1;
-    stopAt = outputSize - 1;
-    for (startAt = zeroCount; startAt < inlen; startAt++) {
-        int carry = in[startAt];
-        for (j = outputSize - 1; (int)j >= 0; j--) {
-            carry += 256 * buffer[j];
-            buffer[j] = carry % 58;
-            carry /= 58;
-
-            if (j <= stopAt - 1 && carry == 0) {
-                break;
-            }
-        }
-        stopAt = j;
+/*
+ * Parses the output as either a 4 or 8-byte unsigned integer.
+ *
+ * Returns the position in buffer after parsing the value.
+ */
+uint8_t* parse_output_value(uint8_t *in, size_t inlen, uint64_t *value) {
+    uint8_t *buf = in;
+    uint64_t tmp = 0;
+    // if first bit is 1, it's 8 bytes long. Otherwise, it's 4
+    bool flag = ((0x80 & in[0]) ? true : false);
+    if (flag) {
+        assert_length(11, inlen);    // value + token_data + script_len
+        tmp = U8BE(in, 0);
+        tmp = (-1)*tmp;
+        buf += 8;
+    } else {
+        tmp = U4BE(in, 0);
+        buf += 4;
     }
+    os_memcpy(value, &tmp, 8);
```

Comment:
Can't you just to `*value = tmp;`?

Reply:
Maybe. To be honest, I didn't test, but I don't see any advantage of one over the other.

Reply:
The first is a simple assignment statement that will be run in a few instructions (if not one), while the `os_memcpy` is a call to a method that will copy byte-by-byte.

Usually `os_memcpy` has a lot of optimizations to do it quickly, but nothing will be faster than a simple assignment.

Here is the libc implementation of memcpy used in Android: https://android.googlesource.com/kernel/lk/+/master/lib/libc/string/memcpy.c

Here is the glibc implementation of memcpy: https://github.com/lattera/glibc/blob/master/string/memcpy.c

Reply:
Done